File: mm-php/src/io/metamessage/jsonc/JsoncNode.php
<?php

namespace io\metamessage\jsonc;

enum JsoncValueType {
    case Unknown;
    case String;
    case Int;
    case Int8;
    case Int16;
    case Int32;
    case Int64;
    case Uint;
    case Uint8;
    case Uint16;
    case Uint32;
    case Uint64;
    case Float32;
    case Float64;
    case Bool;
    case Bytes;
    case BigInt;
    case DateTime;
    case Date;
    case Time;
    case UUID;
    case Decimal;
    case IP;
    case URL;
    case Email;
    case Enum;
    case Array;
    case Struct;
    case Slice;
    case Map;
    case Null;
    case Raw;
}

class JsoncTag {
    public function __construct(
        public string $name = "",
        public bool $isNull = false,
        public JsoncValueType $type = JsoncValueType::Unknown,
        public string $desc = "",
        public bool $raw = false,
        public bool $nullable = false,
        public bool $allowEmpty = false,
        public bool $unique = false,
        public string $defaultValue = "",
        public string $min = "",
        public string $max = "",
        public int $size = 0,
        public string $enum = "",
        public string $pattern = "",
        public string $location = "",
        public int $version = 0,
        public string $mime = "",
        public string $childDesc = "",
        public JsoncValueType $childType = JsoncValueType::Unknown,
        public bool $childRaw = false,
        public bool $childNullable = false,
        public bool $childAllowEmpty = false,
        public bool $childUnique = false,
        public string $childDefault = "",
        public string $childMin = "",
        public string $childMax = "",
        public int $childSize = 0,
        public string $childEnum = "",
        public string $childPattern = "",
        public string $childLocation = "",
        public int $childVersion = 0,
        public string $childMime = "",
        public bool $isInherit = false
    ) {}
}

abstract class JsoncNode {
    public ?JsoncTag $tag = null;
    public string $path = "";
}

class JsoncValue extends JsoncNode {
    public function __construct(
        public mixed $data = null,
        public string $text = "",
        ?JsoncTag $tag = null,
        string $path = ""
    ) {
        $this->tag = $tag;
        $this->path = $path;
    }
}

class JsoncField {
    public function __construct(
        public string $key,
        public ?JsoncNode $value
    ) {}
}

class JsoncObject extends JsoncNode {
    /** @var JsoncField[] */
    public array $fields = [];

    public function __construct(?JsoncTag $tag = null, string $path = "") {
        $this->tag = $tag;
        $this->path = $path;
    }
}

class JsoncArray extends JsoncNode {
    /** @var JsoncNode[] */
    public array $items = [];

    public function __construct(?JsoncTag $tag = null, string $path = "") {
        $this->tag = $tag;
        $this->path = $path;
    }
}

File: mm-kt/src/main/kotlin/io/github/metamessage/jsonc/JsoncNode.kt
package io.github.metamessage.jsonc

sealed class JsoncNode {
    abstract var tag: JsoncTag?
    abstract var path: String
}

data class JsoncValue(
    var data: Any? = null,
    var text: String = "",
    override var tag: JsoncTag? = null,
    override var path: String = ""
) : JsoncNode()

data class JsoncObject(
    var fields: MutableList<JsoncField> = mutableListOf(),
    override var tag: JsoncTag? = null,
    override var path: String = ""
) : JsoncNode()

data class JsoncArray(
    var items: MutableList<JsoncNode> = mutableListOf(),
    override var tag: JsoncTag? = null,
    override var path: String = ""
) : JsoncNode()

data class JsoncField(
    val key: String,
    val value: JsoncNode?
)

data class JsoncTag(
    var name: String = "",
    var isNull: Boolean = false,
    var type: JsoncValueType = JsoncValueType.Unknown,
    var desc: String = "",
    var raw: Boolean = false,
    var nullable: Boolean = false,
    var allowEmpty: Boolean = false,
    var unique: Boolean = false,
    var defaultValue: String = "",
    var min: String = "",
    var max: String = "",
    var size: Int = 0,
    var enum: String = "",
    var pattern: String = "",
    var location: String = "",
    var version: Int = 0,
    var mime: String = "",
    var childDesc: String = "",
    var childType: JsoncValueType = JsoncValueType.Unknown,
    var childRaw: Boolean = false,
    var childNullable: Boolean = false,
    var childAllowEmpty: Boolean = false,
    var childUnique: Boolean = false,
    var childDefault: String = "",
    var childMin: String = "",
    var childMax: String = "",
    var childSize: Int = 0,
    var childEnum: String = "",
    var childPattern: String = "",
    var childLocation: String = "",
    var childVersion: Int = 0,
    var childMime: String = "",
    var isInherit: Boolean = false
)

enum class JsoncValueType {
    Unknown,
    String,
    Int,
    Int8,
    Int16,
    Int32,
    Int64,
    Uint,
    Uint8,
    Uint16,
    Uint32,
    Uint64,
    Float32,
    Float64,
    Bool,
    Bytes,
    BigInt,
    DateTime,
    Date,
    Time,
    UUID,
    Decimal,
    IP,
    URL,
    Email,
    Enum,
    Array,
    Struct,
    Slice,
    Map,
    Null,
    Raw
}